lib/ansible/plugins/strategy/__init__.py

```diff
@@ -481,6 +481,8 @@ def parent_handler_match(target_handler, handler_name):
                                 if original_task.action == 'set_fact':
                                     self._variable_manager.set_nonpersistent_facts(target_host, result_item['ansible_facts'].copy())
                                 else:
+                                    if original_task.action == 'setup':
+                                        target_host._gathered_facts = True
```

Comment:
Running it without this code is enough to get the reproduce case to pass, but that more or less reverts
https://github.com/ansible/ansible/commit/4ac2bafc4baa59d06c937eb8cb9ade0c6ae45a92 

Not really sure how https://github.com/ansible/ansible/commit/4ac2bafc4baa59d06c937eb8cb9ade0c6ae45a92 fixed https://github.com/ansible/ansible/issues/12213 , so just removing it may be a regression. pinging @jimi-c